src/oracle/oracle.cairo

```diff
@@ -258,51 +291,86 @@ mod Oracle {
             data_store: IDataStoreDispatcher,
             event_emitter: IEventEmitterDispatcher,
             params: SetPricesParams,
-        ) { // TODO
+        ) {
+            assert(
+                self.tokens_with_prices.read().len().is_zero(),
+                OracleError::NON_EMPTY_TOKENS_WITH_PRICES
+            );
+
+            self.set_prices_from_price_feeds(data_store, event_emitter, @params.price_feed_tokens);
+            // it is possible for transactions to be executed using just params.priceFeedTokens
+            // in this case if params.tokens is empty, the function can return
+            if params.tokens.len().is_zero() {
+                return;
+            }
+
+            self.set_prices_(data_store, event_emitter, params);
         }
 
-        fn set_primary_price(
-            ref self: ContractState, token: ContractAddress, price: Price,
-        ) { // TODO
+        // Set the primary price
+        // Arguments
+        // * `token` - The token to set the price for.
+        // * `price` - The price value to set to.
+        fn set_primary_price(ref self: ContractState, token: ContractAddress, price: Price,) {
+            self.set_primary_price_(token, price);
         }
 
-        fn clear_all_prices(ref self: ContractState) { // TODO
+        fn clear_all_prices(ref self: ContractState) {
+            let mut length = self.tokens_with_prices.read().len();
+            loop {
+                if length.is_zero() {
+                    break;
+                }
+                let token = self.tokens_with_prices.read().get(0).unwrap();
+                self.remove_primary_price(token);
+                length -= 1;
+            }
         }
 
-        fn get_tokens_with_prices_count(self: @ContractState) -> u128 {
-            // TODO
-            0
+
+        fn get_tokens_with_prices_count(self: @ContractState) -> u32 {
+            self.tokens_with_prices.read().len()
         }
 
         fn get_tokens_with_prices(
-            self: @ContractState, start: u128, end: u128
+            self: @ContractState, start: u32, end: u32
         ) -> Array<ContractAddress> {
-            // TODO
-            ArrayTrait::new()
+            self.tokens_with_prices.read().array()
```

Comment:
Can also get inspiration from data_store PRs given below

Reply:
Refactored fn, now it returns tokens between indexes.